### PR TITLE
Support phone login in modular authentication flow

### PR DESCRIPTION
**Pull request by alexaubry**
### Issues

In this pull request, we add support for phone sign-in in the new authentication coordinator.

### Solutions

- Fix layout issues
- Add steps for add e-mail and password step
- Add steps for e-mail verification

## Notes

We merge this pull request in the WIP branch of #2505. Complete testing will come in a subsequent PR.

### CODE REVIEW

File: Wire-iOS/Sources/AppStateController.swift [hunk 1]
```diff
@@ -206,6 +206,18 @@ extension AppStateController : RegistrationViewControllerDelegate, Authenticatio
         return ZMUserSession.shared()
     }
 
+    func authenticationCoordinatorRequestedSelfUserProfile() -> UserProfile? {
+        guard let session = ZMUserSession.shared() else {
+            return nil
+        }
+
+        return session.userProfile
```
Comment (daehn) [inline Wire-iOS/Sources/AppStateController.swift]: Might be simpler to use optional chaining here and just `return ZMUserSession.shared()?.userProfile`

Comment (daehn) [reply Wire-iOS/Sources/AppStateController.swift]: Also not sure if these accessors should be functions or properties right now, maybe it's fine to make them vars and call them `selfUserProfile`, `selfUser` and `numberOfAccounts`?

File: Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift [hunk 2]
```diff
@@ -203,6 +214,90 @@ extension AuthenticationCoordinator {
...
+
+    @objc func setEmailCredentialsForCurrentUser(_ credentials: ZMEmailCredentials) {
+        guard case let .addEmailAndPassword(_, profile, _) = currentStep else {
+            return
+        }
+
+        transition(to: AuthenticationFlowStep.registerEmailCredentials(credentials))
+        presenter?.showLoadingView = true
+
+        let result = setCredentialsWithProfile(profile, credentials: credentials) == true && SessionManager.shared?.update(credentials: credentials) == true
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift]: I think the `== true` can safely be omitted here.

File: Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift [hunk 3]
```diff
@@ -203,6 +214,90 @@ extension AuthenticationCoordinator {
...
+
+        if !result {
+            let error = NSError(code: .invalidEmail, userInfo: nil)
+            emailUpdateDidFail(error)
+        }
+
+        return
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift]: Can be removed

File: Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift [hunk 4]
```diff
@@ -507,7 +690,7 @@ extension AuthenticationCoordinator {
         }
 
         presenter?.logoEnabled = false
-        transition(to: .addEmailAndPassword, resetStack: true)
+        // transition(to: .addEmailAndPassword(canSkip: false), resetStack: true)
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift]: Can be removed?

File: Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift [hunk 5]
```diff
@@ -507,7 +690,7 @@ extension AuthenticationCoordinator {
...
     }
 
 }
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift]: This file is getting quite big, maybe it makes sense to move some extension into separate files or even thing what can be extracted in other classes (or sub-coordinators) at some point.

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationPersonal/RegistrationPhoneFlowViewController.m [hunk 6]
```diff
@@ -227,7 +227,7 @@ - (void)authenticationDidFail:(NSError *)error
     if (error.code == ZMUserSessionNeedsToRegisterEmailToRegisterClient) {
         if (![self.navigationController.topViewController isKindOfClass:[AddEmailPasswordViewController class]]) {
             AddEmailPasswordViewController *addEmailPasswordViewController = [[AddEmailPasswordViewController alloc] init];
-            addEmailPasswordViewController.formStepDelegate = self;
+//            addEmailPasswordViewController.formStepDelegate = self;
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/RegistrationPersonal/RegistrationPhoneFlowViewController.m]: Can be removed? If not just yet then maybe worth adding a `// TODO`.

File: Wire-iOS/Sources/UserInterface/Registration/RegistrationPersonal/RegistrationViewController.m [hunk 7]
```diff
@@ -151,23 +149,13 @@ - (void)setupNavigationController
 
     if (userSessionErrorCode == ZMUserSessionNeedsToRegisterEmailToRegisterClient) {
         AddEmailPasswordViewController *addEmailPasswordViewController = [[AddEmailPasswordViewController alloc] init];
-        addEmailPasswordViewController.formStepDelegate = self;
+//        addEmailPasswordViewController.formStepDelegate = self;
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Registration/RegistrationPersonal/RegistrationViewController.m]: Same here.

File: Wire-iOS/Sources/UserInterface/Settings/CellDescriptors/SettingsCellDescriptorFactory+Account.swift [hunk 8]
```diff
@@ -116,11 +116,11 @@ extension SettingsCellDescriptorFactory {
                 return ChangeEmailViewController()
              } else {
                 let addEmailController = AddEmailPasswordViewController()
-                addEmailController.showsNavigationBar = false
+//                addEmailController.canSk = false
```
Comment (daehn) [inline Wire-iOS/Sources/UserInterface/Settings/CellDescriptors/SettingsCellDescriptorFactory+Account.swift]: Can be removed?

File: Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift (hunk 5 above)
Comment (alexaubry) [reply Wire-iOS/Sources/UserInterface/Registration/Coordinator/AuthenticationCoordinator.swift]: Agreed, I will look into this now.

**Comment by alexaubry** [comment]
@daehn Updated following your comments. 👍

**Pull request merged**